src/sage/categories/finite_fields.py

```diff
@@ -243,5 +246,227 @@ def _element_of_factored_order(self, F):
+
+            EXAMPLES::
+
+                sage: k.<a> = GF((5, 10))
+                sage: k(2).is_square()
+                True
+                sage: k(2)._cipolla()^2 == k(2)
+                True
+                sage: k.quadratic_non_residue()._cipolla()
+                Traceback (most recent call last):
+                ...
+                ValueError: Element is not a square
+            """
+            parent = self.parent()
+            q = parent.cardinality()
+            if not self.is_square():
+                raise ValueError("Element is not a square")
+            t = parent.random_element()
+            root = t**2 - 4 * self
+            while root.is_square():
+                t = parent.random_element()
+                root = t**2 - 4 * self
+            from sage.rings.polynomial.polynomial_ring import polygen
+            X = polygen(parent)
+            f = X**2 - t*X + self
+            b = pow(X, (q+1)//2, f)
+            return b
+
+        def sqrt(self, all: bool = False, algorithm: str = 'tonelli'):
+            r"""
+            Returns the square root of the element if it exists
+
+            INPUT:
+
+            - ``all`` -- boolean (default: ``False``); whether to return a list of
+              all square roots or just a square root
+
+            - ``algorithm`` -- string (default: 'tonelli'); the algorithm to use
+              among ``'tonelli'``, ``'cipolla'``. Tonelli is typically faster but has
+              a worse worst-case complexity than Cipolla. In particular, if the
+              field cardinality minus 1 is highly divisible by 2 and has a large
+              odd factor then Cipolla may perform better.
+
+            OUTPUT:
+
+            - if ``all=False``, a square root; raises an error if the element is not
+              a square
+
+            - if ``all=True``, a tuple of all distinct square roots. This tuple can have
+              length 0, 1, or 2 depending on how many distinct square roots the
+              element has.
+
+            EXAMPLES::
+
+                sage: S.<x> = GF(5)[]
+                sage: f = S.irreducible_element(20)
+                sage: k.<y> = S.quotient_ring(f)
+                sage: k in Fields()
+                True
+                sage: k(2).is_square()
+                True
+                sage: k(2).sqrt()^2 == k(2)
+                True
+                sage: my_sqrts = k(4).sqrt(all=True)
+                sage: len(k(4).sqrt(all=True))
+                2
+                sage: 2 in my_sqrts
+                True
+                sage: 3 in my_sqrts
+                True
+                sage: k.quadratic_non_residue().sqrt()
+                Traceback (most recent call last):
+                ...
+                ValueError: Element is not a square
+                sage: k.quadratic_non_residue().sqrt(all=True)
+                ()
+
+            ALGORITHM:
+
+            The algorithms used come from chapter 6 of [BS1996]_.
```

Comment:
```suggestion
            The algorithms used come from chapter 7 of [BS1996]_.
```

I don't know why I suggested to change this from chapter 8 to 6 earlier. Tonelli and Cipolla are definitely in chapter 7.